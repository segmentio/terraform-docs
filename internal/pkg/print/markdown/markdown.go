package markdown

import (
	"bytes"
	"fmt"
	"strings"

	"github.com/segmentio/terraform-docs/internal/pkg/doc"
	"github.com/segmentio/terraform-docs/internal/pkg/print"
	"github.com/segmentio/terraform-docs/internal/pkg/settings"
)

// Print prints a document as markdown.
func Print(document *doc.Doc, settings settings.Settings) (string, error) {
	var buffer bytes.Buffer

	if document.HasComment() {
		printComment(&buffer, document.Comment, settings)
	}

	if document.HasInputs() {
		if settings.Has(print.WithSorting) {
			doc.SortInputsByName(document.Inputs)
		}

		printInputs(&buffer, document.Inputs, settings)
	}

	if document.HasOutputs() {
		if settings.Has(print.WithSorting) {
			doc.SortOutputsByName(document.Outputs)
		}

		printOutputs(&buffer, document.Outputs, settings)
	}

	return buffer.String(), nil
}

func getInputDefaultValue(input *doc.Input, settings settings.Settings) string {
	var result = "-"

	if input.HasDefault() {
		result = fmt.Sprintf("`%s`", print.GetPrintableValue(input.Default, settings))
	}

	return result
}

func getInputDescription(input *doc.Input) string {
	var result = "-"

	if input.HasDescription() {
		result = input.Description
	}

	return result
}

func getOutputDescription(output *doc.Output) string {
	var result = "-"

	if output.HasDescription() {
		result = output.Description
	}

	return result
}

func printComment(buffer *bytes.Buffer, comment string, settings settings.Settings) {
	buffer.WriteString(fmt.Sprintf("%s\n", comment))
}

func printInputs(buffer *bytes.Buffer, inputs []doc.Input, settings settings.Settings) {
	buffer.WriteString("\n## Inputs\n\n")
	buffer.WriteString("| Name | Description | Type | Default |")

	if settings.Has(print.WithRequired) {
		buffer.WriteString(" Required |\n")
	} else {
		buffer.WriteString("\n")
	}

	buffer.WriteString("|------|-------------|:----:|:-----:|")

	if settings.Has(print.WithRequired) {
		buffer.WriteString(":-----:|\n")
	} else {
		buffer.WriteString("\n")
	}

	for _, input := range inputs {
		buffer.WriteString(
			fmt.Sprintf("| %s | %s | %s | %s |",
				input.Name,
				prepareDescriptionForMarkdown(getInputDescription(&input)),
				input.Type,
				getInputDefaultValue(&input, settings)))

		if settings.Has(print.WithRequired) {
			buffer.WriteString(fmt.Sprintf(" %v |\n", printIsInputRequired(&input)))
		} else {
			buffer.WriteString("\n")
		}
	}
}

func printIsInputRequired(input *doc.Input) string {
	if input.IsRequired() {
		return "yes"
	}

	return "no"
}

func printOutputs(buffer *bytes.Buffer, outputs []doc.Output, settings settings.Settings) {
	buffer.WriteString("\n## Outputs\n\n")
	buffer.WriteString("| Name | Description |\n")
	buffer.WriteString("|------|-------------|\n")

	for _, output := range outputs {
		buffer.WriteString(
			fmt.Sprintf("| %s | %s |\n",
				output.Name,
				prepareDescriptionForMarkdown(getOutputDescription(&output))))
	}
}

func prepareDescriptionForMarkdown(s string) string {
	// Convert double newlines to <br><br>.
	s = strings.Replace(
		strings.TrimSpace(s),
		"\n\n",
		"<br><br>",
		-1)

	// Convert single newline to space.
	return strings.Replace(s, "\n", " ", -1)
}
