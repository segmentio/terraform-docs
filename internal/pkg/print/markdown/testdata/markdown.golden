Usage:

module "foo" {
  source = "github.com/foo/bar"

  id   = "1234567890"
  name = "baz"

  zones = ["us-east-1", "us-west-1"]

  tags = {
    Name         = "baz"
    Created-By   = "first.last@email.com"
    Date-Created = "20180101"
  }
}

## Inputs

| Name | Description | Type | Default |
|------|-------------|:----:|:-----:|
| string-2 | It's string number two. | string | - |
| string-1 | It's string number one. | string | `bar` |
| map-3 | - | map | `<map>` |
| map-2 | It's map number two. | map | - |
| map-1 | It's map number one. | map | `<map>` |
| list-3 | - | list | `<list>` |
| list-2 | It's list number two. | list | - |
| list-1 | It's list number one. | list | `<list>` |

## Outputs

| Name | Description |
|------|-------------|
| output-2 | It's output number two. |
| output-1 | It's output number one. |
